stl/inc/ranges

```diff
@@ -6821,6 +6821,682 @@ namespace ranges {
+            const auto _Temp = *this;
+            --*this;
+
+            return _Temp;
+        }
+
+        constexpr _Iterator& operator+=(const difference_type _Off) noexcept(_Is_plus_equals_nothrow) // strengthened
+            requires _Zip::_All_random_access<_IsConst, _ViewTypes...> {
+            if constexpr (is_arithmetic_v<difference_type>) {
+                _Zip::_Tuple_for_each(
+                    [&_Off]<class _IteratorType>(
+                        _IteratorType& _Itr) { _Itr += static_cast<iter_difference_t<_IteratorType>>(_Off); },
+                    _Current);
+            } else {
+                _Zip::_Tuple_for_each([&_Off]<class _IteratorType>(
+                                          _IteratorType& _Itr) { _Itr += iter_difference_t<_IteratorType>{_Off}; },
+                    _Current);
+            }
+
+            return *this;
+        }
+
+        constexpr _Iterator& operator-=(const difference_type _Off) noexcept(_Is_minus_equals_nothrow) // strengthened
+            requires _Zip::_All_random_access<_IsConst, _ViewTypes...> {
+            if constexpr (is_arithmetic_v<difference_type>) {
+                _Zip::_Tuple_for_each(
+                    [&_Off]<class _IteratorType>(
+                        _IteratorType& _Itr) { _Itr -= static_cast<iter_difference_t<_IteratorType>>(_Off); },
+                    _Current);
+            } else {
+                _Zip::_Tuple_for_each([&_Off]<class _IteratorType>(
+                                          _IteratorType& _Itr) { _Itr -= iter_difference_t<_IteratorType>{_Off}; },
+                    _Current);
+            }
+
+            return *this;
+        }
+
+        _NODISCARD constexpr auto operator[](const difference_type _Where) const
+            noexcept(_Is_brace_indexing_nothrow) // strengthened
+            requires _Zip::_All_random_access<_IsConst, _ViewTypes...> {
+            if constexpr (is_arithmetic_v<difference_type>) {
+                return _Zip::_Tuple_transform(
+                    [&_Where]<class _IteratorType>(_IteratorType& _Itr) -> decltype(auto) {
+                        return _Itr[static_cast<iter_difference_t<_IteratorType>>(_Where)];
+                    },
+                    _Current);
+            } else {
+                return _Zip::_Tuple_transform(
+                    [&_Where]<class _IteratorType>(_IteratorType& _Itr) -> decltype(auto) {
+                        return _Itr[iter_difference_t<_IteratorType>{_Where}];
+                    },
+                    _Current);
+            }
+        }
+
+        _NODISCARD_FRIEND constexpr bool operator==(const _Iterator& _Lhs, const _Iterator& _Rhs) noexcept(
+            _Is_equality_comparison_nothrow) // strengthened
+            requires(equality_comparable<iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>&&...) {
+            if constexpr (_Zip::_All_bidirectional<_IsConst, _ViewTypes...>) {

```

Comment:
`get<1>(_Lhs._Current) == get<1>(_Rhs._Current)` is not guaranteed to be equal when `get<1>(_Lhs._Current) == _STD get<1>(_Rhs._Current)` is equal, since an exception may have been thrown while creating the second iterator.

Reply:
@hewillk if an exception is thrown, then the iterator has a "singular value" for which `==` is undefined behavior.

Reply:
Let's get rid of the exception case.
My first question is, why we cannot check `_STD get<0>(_Lhs._Current) == _STD get<0>(_Rhs._Current)` when `_Zip::_All_bidirectional<_IsConst, _ViewTypes...>` is modeled? 
Why must `_All_random_access<_IsConst, _ViewTypes...>` be modeled?

Reply:
Mmh, that's a good point. I don't believe it must be. For `_Is_common && !_All_random_access`, you need to make sure that you're not comparing against the end iterator (hence the need for a `any_of`), but for `!_Is_common || _All_random_access` iterators are always derived from the `begin()` iterator and all of the underlying iterators have been `++`'d the same number of times.

Reply:
After lots of contemplation, I think I finally understand what you are getting at here. If we look at how `std::ranges::zip_view::end()` is defined, it returns either an `_Iterator` or an `_Sentinel` instance. The case we need to watch out for is when the function returns an `_Iterator` containing the value of `std::ranges::end(_View)` for every view in `_Views`. Only when the expression `_Zip_is_common<_Maybe_const<_IsConst, _ViewTypes>...> && !_All_random_access<_IsConst, _ViewTypes...>` is true does this function return that dangerous value.

The reason this is dangerous is because then `operator==(const _Iterator& _Lhs, const _Iterator& _Rhs)` will allow the comparison between valid iterators and end iterators which are never actually used to denote the end of the zipped range. Indeed, this very issue was discussed in the paper P2321R2 at [this section](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2321r2.html#when-is-zip_view-a-common_range). We only want to do these comparisons when we know that the `_Iterator` is in the valid domain for `operator==`.

With this epiphany, it seems that we should be able to optimize the function as follows (proper formatting not included):

```C++
_NODISCARD_FRIEND constexpr bool operator==(const _Iterator& _Lhs, const _Iterator& _Rhs) [. . .] 
{
  if constexpr (!_Zip_is_common<_Maybe_const<_IsConst, _ViewTypes>...> || _All_random_access<_IsConst, _ViewTypes...>)
    return (_STD get<0>(_Lhs._Current) == _STD get<0>(_Rhs._Current));

  else if constexpr (_All_bidirectional<_IsConst, _ViewTypes...>)
    return (_Lhs._Current == _Rhs._Current);

  else
    return _Zip_evaluate_iterator_sentinel_equality(_Lhs._Current, _Rhs._Current);
}
```

That's basically what @strega-nil-ms said in her previous comment here, but I just wanted to re-affirm and prove that what she was saying is, in fact, correct.